source/docs/software/advanced-controls/state-space/state-space-flywheel-walkthrough.rst

```diff
@@ -0,0 +1,250 @@
+
+.. todo:: do we need to elaborate on this^ more?
+
+Linear-Quadratic Regulators and Plant Inversion Feedforward
+-----------------------------------------------------------
+
+:ref:`docs/software/advanced-controls/state-space/state-space-intro:The Linear-Quadratic Regulator` finds a feedback controller to drive our flywheel :term:`system` to its :term:`reference`. Because our flywheel has just one state, the control law picked by our LQR will be in the form :math:`\mathbf{u = K (r - x)}` where :math:`\mathbf{K}` is a 1x1 matrix; in other words, the control law picked by LQR is simply a proportional controller, or a PID controller with only a P gain. This gain is chosen by our LQR based on the state excursion and control efforts we pass it. More on tuning LQR controllers can be found in the :ref:`LQR application example <docs/software/advanced-controls/state-space/state-space-intro:LQR: example application>`.
+
+Much like ``SimpleMotorFeedforward`` can be used to generate feedforward voltage inputs given kS, kV, and kA constants, the Plant Inversion Feedforward class generate :ref:`feedforward <docs/software/advanced-controls/state-space/state-space-intro:Visualizing Feedforward>` voltage inputs given a state-space system. The voltage commands generated by the ``LinearSystemLoop`` class are the sum of the feedforward and feedback inputs.
+
+.. tabs::
+
+   .. group-tab:: Java
+
+      .. remoteliteralinclude:: https://raw.githubusercontent.com/wpilibsuite/allwpilib/master/wpilibjExamples/src/main/java/edu/wpi/first/wpilibj/examples/statespaceflywheel/Robot.java
+         :language: java
+         :lines: 63-72
+         :linenos:
+         :lineno-start: 63
+
+   .. group-tab:: C++
+
+      .. remoteliteralinclude:: https://raw.githubusercontent.com/wpilibsuite/allwpilib/master/wpilibcExamples/src/main/cpp/examples/StateSpaceFlywheelSysId/cpp/Robot.cpp
+         :language: cpp
+         :lines: 15-16
+         :linenos:
+         :lineno-start: 15
+
+      .. remoteliteralinclude:: https://raw.githubusercontent.com/wpilibsuite/allwpilib/master/wpilibcExamples/src/main/cpp/examples/StateSpaceFlywheel/cpp/Robot.cpp
+         :language: cpp
+         :lines: 59-79
+         :linenos:
+         :lineno-start: 59
+
+Bringing it All Together: LinearSystemLoop
+------------------------------------------
+
+LinearSystemLoop combines our system, controller, and observer that we created earlier. The constructor shown will also instantiate a ``PlantInversionFeedforward``.
```

Comment:
The code snippets show no feedforward being passed to Java (I'm assuming it's being created internally), but the feedforward is being passed in C++. Is this intentional?

Reply:
That might be an issue with the wpilib example classes. in java, we use the ff created internally.

Reply:
Yeah, pending fix in https://github.com/wpilibsuite/allwpilib/pull/2755